### ISSUE TITLE
Keras Code Example for image similarity estimation: model's final layer sigmoid activation is inconsistent with contrastive loss (with integer margin)

### ISSUE THREAD

**Issue by chanangaza** [keras-team/keras-io#2230]
### Issue Type

Performance

### Source

binary

### Keras Version

3.13

### Custom Code

Yes

### Python version

3.12

### Current Behavior?

in [Image similarity estimation using a Siamese Network with a contrastive loss](https://keras.io/examples/vision/siamese_contrastive/):

Shouldn't we use Binary-Cross-Entropy Loss given the final dense classification head with Sigmoid activation? the integer 'margin' param makes no sense... either remove final dense layer and have model output raw Eucleadian distance against the contrastive-loss (integer margin) or use the Binary-Cross-Entropy loss (but then this is not an example of contrastive loss)...

### Standalone code to reproduce the issue or tutorial link

```shell
### from example...
# Provided two tensors t1 and t2
# Euclidean distance = sqrt(sum(square(t1-t2)))
def euclidean_distance(vects):
    """Find the Euclidean distance between two vectors.

    Arguments:
        vects: List containing two tensors of same length.

    Returns:
        Tensor containing euclidean distance
        (as floating point value) between vectors.
    """

    x, y = vects
    sum_square = ops.sum(ops.square(x - y), axis=1, keepdims=True)
    return ops.sqrt(ops.maximum(sum_square, keras.backend.epsilon()))


input = keras.layers.Input((28, 28, 1))
x = keras.layers.BatchNormalization()(input)
x = keras.layers.Conv2D(4, (5, 5), activation="tanh")(x)
x = keras.layers.AveragePooling2D(pool_size=(2, 2))(x)
x = keras.layers.Conv2D(16, (5, 5), activation="tanh")(x)
x = keras.layers.AveragePooling2D(pool_size=(2, 2))(x)
x = keras.layers.Flatten()(x)

x = keras.layers.BatchNormalization()(x)
x = keras.layers.Dense(10, activation="tanh")(x)
embedding_network = keras.Model(input, x)


input_1 = keras.layers.Input((28, 28, 1))
input_2 = keras.layers.Input((28, 28, 1))

# As mentioned above, Siamese Network share weights between
# tower networks (sister networks). To allow this, we will use
# same embedding network for both tower networks.
tower_1 = embedding_network(input_1)
tower_2 = embedding_network(input_2)

merge_layer = keras.layers.Lambda(euclidean_distance, output_shape=(1,))(
    [tower_1, tower_2]
)
normal_layer = keras.layers.BatchNormalization()(merge_layer)
output_layer = keras.layers.Dense(1, activation="sigmoid")(normal_layer)
siamese = keras.Model(inputs=[input_1, input_2], outputs=output_layer)
Define the contrastive Loss
def loss(margin=1):
    """Provides 'contrastive_loss' an enclosing scope with variable 'margin'.

    Arguments:
        margin: Integer, defines the baseline for distance for which pairs
                should be classified as dissimilar. - (default is 1).

    Returns:
        'contrastive_loss' function with data ('margin') attached.
    """

    # Contrastive loss = mean( (1-true_value) * square(prediction) +
    #                         true_value * square( max(margin-prediction, 0) ))
    def contrastive_loss(y_true, y_pred):
        """Calculates the contrastive loss.

        Arguments:
            y_true: List of labels, each label is of type float32.
            y_pred: List of predictions of same length as of y_true,
                    each label is of type float32.

        Returns:
            A tensor containing contrastive loss as floating point value.
        """

        square_pred = ops.square(y_pred)
        margin_square = ops.square(ops.maximum(margin - (y_pred), 0))
        return ops.mean((1 - y_true) * square_pred + (y_true) * margin_square)

    return contrastive_loss
```

### Relevant log output

```shell

```

**Comment by maitry63** [comment]
Hi @chanangaza, Thank you for catching this issue.  You're correct the sigmoid layer conflicts with the contrastive loss, which expects raw Euclidean distances.
We can remove the final classification layers so we output raw distance.

Example:

```
### from example...
# Provided two tensors t1 and t2
# Euclidean distance = sqrt(sum(square(t1-t2)))

def euclidean_distance(vects):
    """Find the Euclidean distance between two vectors.

    Arguments:
        vects: List containing two tensors of same length.

    Returns:
        Tensor containing euclidean distance
        (as floating point value) between vectors.
    """
    x, y = vects
    sum_square = ops.sum(ops.square(x - y), axis=1, keepdims=True)
    return ops.sqrt(ops.maximum(sum_square, keras.backend.epsilon()))

# embedding network
def create_embedding_network():
    input = keras.layers.Input((28, 28, 1))
    x = keras.layers.BatchNormalization()(input)
    x = keras.layers.Conv2D(4, (5, 5), activation="tanh")(x)
    x = keras.layers.AveragePooling2D(pool_size=(2, 2))(x)
    x = keras.layers.Conv2D(16, (5, 5), activation="tanh")(x)
    x = keras.layers.AveragePooling2D(pool_size=(2, 2))(x)
    x = keras.layers.Flatten()(x)
    x = keras.layers.BatchNormalization()(x)
    x = keras.layers.Dense(10, activation="tanh")(x)
    return keras.Model(input, x, name="embedding")

embedding_network = create_embedding_network()

# Siamese model outputs 
input_1 = keras.layers.Input((28, 28, 1))
input_2 = keras.layers.Input((28, 28, 1))
tower_1 = embedding_network(input_1)
tower_2 = embedding_network(input_2)
distance = keras.layers.Lambda(euclidean_distance, output_shape=(1,))([tower_1, tower_2])
siamese_model = keras.Model(inputs=[input_1, input_2], outputs=distance)

# Define the contrastive Loss
def loss(margin=1):
    """Provides 'contrastive_loss' an enclosing scope with variable 'margin'.

    Arguments:
        margin: Integer, defines the baseline for distance for which pairs
                should be classified as dissimilar. - (default is 1).

    Returns:
        'contrastive_loss' function with data ('margin') attached.
    """

    # Contrastive loss = mean( (1-true_value) * square(prediction) +
    #                         true_value * square( max(margin-prediction, 0) ))

    def contrastive_loss(y_true, y_pred):
        """Calculates the contrastive loss.

        Arguments:
            y_true: List of labels, each label is of type float32.
            y_pred: List of predictions of same length as of y_true,
                    each label is of type float32.

        Returns:
            A tensor containing contrastive loss as floating point value.
        """

        square_pred = ops.square(y_pred)
        margin_square = ops.square(ops.maximum(margin - (y_pred), 0))
        return ops.mean((1 - y_true) * square_pred + (y_true) * margin_square)

    return contrastive_loss
```
Will create a PR to update the official keras-io example with this fix.

**Comment by maitry63** [comment]
Hi @chanangaza, regarding your query, the PR has been merged into the official Keras-io repository. Kindly check. Thank you

**Comment by chanangaza** [comment]
Many thanks for your quick response! 😊

**Comment by sachinprasadhs** [comment]
Closing the issue, since the PR is merged. Thanks
